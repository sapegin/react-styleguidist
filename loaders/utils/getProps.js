'use strict';

const path = require('path');
const reactDocs = require('react-docgen');
const highlightCode = require('./highlightCode');
const removeDoclets = require('./removeDoclets');
const requireIt = require('./requireIt');

const examplesLoader = path.resolve(__dirname, '../examples-loader.js');

/**
 * Replace example doclet with a require statement AST.
 *
 * @param {object} doc
 * @returns {object}
 */
module.exports = function getProps(doc) {
	if (doc.description) {
		// Read doclets from the description and remove them
		doc.doclets = reactDocs.utils.docblock.getDoclets(doc.description);
		doc.description = removeDoclets(doc.description);
		doc.description = highlightCode(doc.description);

		if (doc.doclets.example) {
			doc.example = requireIt(`!!${examplesLoader}!${doc.doclets.example}`);
			delete doc.doclets.example;
		}
	}
	else {
		doc.doclets = {};
	}

	return doc;
};
