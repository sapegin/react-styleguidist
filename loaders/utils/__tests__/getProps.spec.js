jest.mock('../requireIt');

import getProps from '../getProps';

it('getProps() should return an object for props', () => {
	const result = getProps({
		description: 'The only true button.',
		methods: [],
		props: {
			children: {
				type: {},
				required: true,
				description: 'Button label.',
			},
			color: {
				type: {},
				required: false,
				description: '',
			},
		},
	});

	expect(result).toMatchSnapshot();
});

it('getProps() should return an object for props without description', () => {
	const result = getProps({
		props: {
			children: {
				type: {},
				required: true,
				description: 'Button label.',
			},
		},
	});

	expect(result).toMatchSnapshot();
});

it('getProps() should return an object for props with doclets', () => {
	const result = getProps({
		description: `
The only true button.

@example example.md
`,
	});

	expect(result).toMatchSnapshot();
});
