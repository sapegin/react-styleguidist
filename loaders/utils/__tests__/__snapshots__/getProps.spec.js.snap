exports[`test getProps() should return an object for props 1`] = `
Object {
  "description": "The only true button.
",
  "doclets": Object {},
  "methods": Array [],
  "props": Object {
    "children": Object {
      "description": "Button label.",
      "required": true,
      "type": Object {},
    },
    "color": Object {
      "description": "",
      "required": false,
      "type": Object {},
    },
  },
}
`;

exports[`test getProps() should return an object for props with doclets 1`] = `
Object {
  "description": "The only true button.
",
  "doclets": Object {},
  "example": RequireStatement {
    "filepath": "!!<rootDir>/loaders/examples-loader.js!example.md
",
  },
}
`;

exports[`test getProps() should return an object for props without description 1`] = `
Object {
  "doclets": Object {},
  "props": Object {
    "children": Object {
      "description": "Button label.",
      "required": true,
      "type": Object {},
    },
  },
}
`;
